include/vapor/unique_ptr_cache.hpp

```diff
@@ -9,130 +9,135 @@
         }
-        else
-            return nullptr;
     }
 
-    //
-    // Inserts to the head of the cache a key and a raw pointer pointing to a BigObj.
-    // Upon success, this class takes ownership of the BigObj and the old raw pointer 
-    // shall not be used anymore.
-    // In the case of the key already exists, the new BigObj takes place of the old 
-    // BigObj, while the old BigObj is properly destroyed.
-    // (Pass by value and move idiom on Key)
-    //
     void insert( Key key, const BigObj* ptr )
     {
-        // Remove the old pair if the same key already exists.
-        // (Should use const iterator here, but gcc-4.8 on CentOS7 doesn't support...)
-        for( auto it = m_list.begin(); it != m_list.end(); ++it )
-        { 
-            if( it->first == key )
-            {
-                m_list.erase( it );
-                break;  // There's at most one existing key, so OK to break here.
+        const std::lock_guard<std::mutex> lock( element_array_mutex );
+
+        auto it = std::find_if( element_array.begin(), element_array.end(), 
+                                [&key](element_type& e){return e.first == key;} );
+
+        if( it == element_array.end() ) {           // This key does not exist
+            if( current_size < CacheCapacity ) {    // This cache is not full
+                element_array[current_size].first = std::move(key);
+                element_array[current_size].second.reset(ptr);
+                std::rotate( element_array.begin(), element_array.begin() + current_size,
+                             element_array.begin() + current_size + 1 );
+                current_size++;
+            }
+            else {                                  // The cache is full!
+                element_array.back().first = std::move(key);
+                element_array.back().second.reset(ptr);
+                std::rotate( element_array.begin(), element_array.end() - 1, element_array.end() );
             }
         }
-
-        // Should have used make_unique<> in C++14. GCC-4.8 in CentOS7 prevents it as of 2019.
-        std::unique_ptr<const BigObj> tmp( ptr );
-
-        // Create a new pair at the front of the list
-        m_list.emplace_front( std::move(key), std::move(tmp) );
-
-        if( m_list.size() > m_max_size )
-            m_list.pop_back();
+        else {                                      // This key does exist. 
+            it->second.reset(ptr);
+            auto tmp = std::move(*it);
+            std::move_backward( element_array.begin(), it, it + 1 );
+            element_array.front() = std::move(tmp);
+        }
     }
 
 private:
-    using list_type = std::list< std::pair<Key, std::unique_ptr<const BigObj>> >;
-    mutable list_type   m_list;
-    const   size_t      m_max_size;
+    using element_type = std::pair<Key, std::unique_ptr<const BigObj>>;
 
-};  // end of class unique_ptr_cache
+    std::array< element_type, CacheCapacity >   element_array;
```

Comment:
These member variables should follow naming conventions (precede with an "_").

Reply:
This is addressed.